File: samples/VB .NET Samples/ShowInfoAndSaveImage/ShowInfoAndSaveImage/Form1.vb
Imports Dynamsoft.Core
Imports Dynamsoft.Core.Enums
Imports Dynamsoft.TWAIN
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports Dynamsoft.TWAIN.Interface
Imports Dynamsoft.PDF
Imports System.Runtime.InteropServices
Imports System.IO

Partial Public Class Form1
    Inherits Form
    Implements IAcquireCallback
    Implements ISave
    Private sImageType As EnumImageFileFormat
    Private m_TwainManager As TwainManager = Nothing
    Private m_ImageCore As ImageCore = Nothing
    Private m_PDFCreator As Dynamsoft.PDF.PDFCreator = Nothing
    Private m_StrProductKey As String
    Public Sub New()
        InitializeComponent()
        m_StrProductKey = "t0068UwAAAEQABDxqjGfgEzhVYureL0kGxugcsvIqCDGTPTsR5nLaQsNupIc17Y5vpMZAWBDsd6Xw3NMYzdHlHwiKUrfe/cU="
        m_TwainManager = New TwainManager(m_StrProductKey)
        m_ImageCore = New ImageCore()
        dsViewer1.Bind(m_ImageCore)
        dsViewer1.SetViewMode(1, 1)
        BMPradio.Checked = True
        sImageType = 0
        MultiTIFF.Enabled = False
        MultiPDF.Enabled = False
        m_PDFCreator = New PDFCreator(m_StrProductKey)

        m_RefreshInfo = New RefreshInfo(AddressOf ShowImageInfo)
    End Sub



    Private Delegate Sub RefreshInfo(strImageXResolution As Integer, strImageYResolution As Integer, strImageWidth As Integer, strImageLength As Integer, strImageBitsPerixel As String, strImagePixelType As String, _
        strImageLayoutFrameLeft As String, strImageLayoutFrameTop As String, strImageLayoutFrameRight As String, strImageLayoutFrameBottom As String, strImageLayoutDocumentNumber As String, strImageLayoutPageNumber As String, _
        strImageLayoutFrameNumber As String)


    Private m_RefreshInfo As RefreshInfo

    Public Function GetImageSize() As String
        Dim ImageSize As String = ""
        If m_ImageCore.ImageBuffer.HowManyImagesInBuffer <> 0 Then
            ImageSize = m_ImageCore.IO.GetImageSizeWithSpecifiedType(m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer, sImageType).ToString()
        End If
        Return ImageSize
    End Function

    Private Sub btnScan_Click(sender As Object, e As EventArgs)

        Dim tempList As New List(Of String)()
        For i As Integer = 0 To m_TwainManager.SourceCount - 1
            tempList.Add(m_TwainManager.SourceNameItems(CShort(i)))
        Next
        Dim tempSourceListWrapper As New SourceListWrapper(tempList)

        Dim iSelectIndex As Integer = tempSourceListWrapper.SelectSource()
        If iSelectIndex = -1 Then
            Return
        Else
            m_TwainManager.IfDisableSourceAfterAcquire = True
            m_TwainManager.IfShowUI = True
            m_TwainManager.SelectSourceByIndex(CShort(iSelectIndex))
            m_TwainManager.AcquireImage(TryCast(Me, IAcquireCallback))
        End If
    End Sub

    Private Sub TIFFradio_CheckedChanged(sender As Object, e As EventArgs)
        MultiTIFF.Enabled = True
        MultiPDF.Enabled = False
        sImageType = EnumImageFileFormat.WEBTW_TIF
        txtFileSize.Text = GetImageSize()
    End Sub

    Private m_PDFFileBytes As Byte() = Nothing
    Private Sub PDFradio_CheckedChanged(sender As Object, e As EventArgs)
        MultiTIFF.Enabled = False
        MultiPDF.Enabled = True
        m_PDFFileBytes = Nothing
        m_PDFFileBytes = m_PDFCreator.SaveAsBytes(TryCast(Me, ISave))
        txtFileSize.Text = "0"
        If m_PDFFileBytes IsNot Nothing Then
            txtFileSize.Text = m_PDFFileBytes.Length.ToString()
        End If

    End Sub

    Private Sub BMPradio_CheckedChanged(sender As Object, e As EventArgs)
        MultiTIFF.Enabled = False
        MultiPDF.Enabled = False
        sImageType = EnumImageFileFormat.WEBTW_BMP
        txtFileSize.Text = GetImageSize()

    End Sub

    Private Sub JPEGradio_CheckedChanged(sender As Object, e As EventArgs)
        MultiTIFF.Enabled = False
        MultiPDF.Enabled = False
        sImageType = EnumImageFileFormat.WEBTW_JPG
        txtFileSize.Text = GetImageSize()
    End Sub

    Private Sub PNGradio_CheckedChanged(sender As Object, e As EventArgs)
        MultiTIFF.Enabled = False
        MultiPDF.Enabled = False
        sImageType = EnumImageFileFormat.WEBTW_PNG
        txtFileSize.Text = GetImageSize()
    End Sub

    Private Sub BMPradio_MouseClick(sender As Object, e As MouseEventArgs)
        MultiTIFF.Enabled = False
        MultiPDF.Enabled = False
        sImageType = 0
        txtFileSize.Text = GetImageSize()
    End Sub

    Private Sub btnSave_Click(sender As Object, e As EventArgs)
        Try
            If m_ImageCore.ImageBuffer.HowManyImagesInBuffer > 0 Then
                Dim strFile As String = ""
                If BMPradio.Checked = True Then
                    dlgFileSave.Filter = "BMP File (*.bmp)|*.bmp"
                ElseIf JPEGradio.Checked = True Then
                    dlgFileSave.Filter = "JPEG File (*.jpg)|*.jpg"
                ElseIf PNGradio.Checked = True Then
                    dlgFileSave.Filter = "PNG File (*.png)|*.png"
                ElseIf TIFFradio.Checked = True Then
                    dlgFileSave.Filter = "TIFF File (*.tif)|*.tif"
                ElseIf PDFradio.Checked = True Then
                    dlgFileSave.Filter = "PDF File (*.pdf)|*.pdf"
                End If
                dlgFileSave.InitialDirectory = System.IO.Directory.GetCurrentDirectory()
                dlgFileSave.FileName = ""
                If dlgFileSave.ShowDialog() = DialogResult.OK Then
                    strFile = dlgFileSave.FileName


                    If BMPradio.Checked = True Then
                        m_ImageCore.IO.SaveAsBMP(strFile, m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer)
                    ElseIf JPEGradio.Checked = True Then

                        m_ImageCore.IO.SaveAsJPEG(strFile, m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer)
                    ElseIf PNGradio.Checked = True Then

                        m_ImageCore.IO.SaveAsPNG(strFile, m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer)
                    ElseIf TIFFradio.Checked = True Then
                        Dim tempImageIndex As New List(Of Short)()
                        If MultiTIFF.Checked = True Then
                            For sIndex As Short = 0 To m_ImageCore.ImageBuffer.HowManyImagesInBuffer - 1
                                tempImageIndex.Add(sIndex)
                            Next
                        Else
                            tempImageIndex.Add(m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer)
                        End If

                        m_ImageCore.IO.SaveAsTIFF(strFile, tempImageIndex)
                    ElseIf PDFradio.Checked = True Then
                        m_PDFCreator.Save(TryCast(Me, ISave), strFile)
                    End If
                End If
            End If
        Catch ex As Exception
            MessageBox.Show(ex.Message)
        End Try

    End Sub

    Private Sub btnRemove_Click(sender As Object, e As EventArgs)
        If m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer > -1 Then
            m_ImageCore.ImageBuffer.RemoveImage(m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer)
            btnScan.Enabled = True
        End If
    End Sub

    Private Sub Clear()
        m_ImageCore.ImageBuffer.RemoveAllImages()
        edtXResolution.Text = ""
        edtYResolution.Text = ""
        edtWidth.Text = ""
        edtLength.Text = ""
        edtBitsPerPixel.Text = ""
        edtPixelType.Text = ""

        edtFrameLeft.Text = ""
        edtFrameTop.Text = ""
        edtFrameRight.Text = ""
        edtFrameBottom.Text = ""
        edtDocNumber.Text = ""
        edtPageNumber.Text = ""
        edtFrameNumber.Text = ""

        txtFileSize.Text = GetImageSize()
    End Sub

    Private Sub dynamicDotNetTwain_OnPostTransfer()
        If m_ImageCore.ImageBuffer.HowManyImagesInBuffer >= m_ImageCore.ImageBuffer.MaxImagesInBuffer Then
            btnScan.Enabled = False
        End If
    End Sub

    Private Sub ShowImageInfo(strImageXResolution As Integer, strImageYResolution As Integer, strImageWidth As Integer, strImageLength As Integer, strImageBitsPerixel As String, strImagePixelType As String, _
        strImageLayoutFrameLeft As String, strImageLayoutFrameTop As String, strImageLayoutFrameRight As String, strImageLayoutFrameBottom As String, strDocumentNumber As String, strPageNumber As String, _
        strFrameNumber As String)
        edtXResolution.Text = strImageXResolution.ToString()
        edtYResolution.Text = strImageYResolution.ToString()
        edtWidth.Text = strImageWidth.ToString()
        edtLength.Text = strImageLength.ToString()
        edtBitsPerPixel.Text = strImageLength.ToString()
        edtPixelType.Text = strImagePixelType.ToString()
        edtFrameLeft.Text = strImageLayoutFrameLeft.ToString()
        edtFrameTop.Text = strImageLayoutFrameTop.ToString()
        edtFrameRight.Text = strImageLayoutFrameRight.ToString()
        edtFrameBottom.Text = strImageLayoutFrameBottom.ToString()


        edtDocNumber.Text = strDocumentNumber
        edtPageNumber.Text = strPageNumber
        edtFrameNumber.Text = strFrameNumber
        edtPageNumber.Text = m_TwainManager.ImageLayoutPageNumber.ToString()
        edtFrameNumber.Text = m_TwainManager.ImageLayoutFrameNumber.ToString()

        txtFileSize.Text = GetImageSize()
    End Sub

    Private Sub dynamicDotNetTwain_OnPreAllTransfers()
        Clear()
    End Sub

    Public Sub LoadConvertResult(result As ConvertResult)
    End Sub

    Public Sub OnPostAllTransfers() Implements IAcquireCallback.OnPostAllTransfers
        Me.BeginInvoke(m_RefreshInfo, CInt(Math.Truncate(m_TwainManager.ImageXResolution)), CInt(Math.Truncate(m_TwainManager.ImageYResolution)), CInt(m_TwainManager.ImageWidth), CInt(m_TwainManager.ImageLength), m_TwainManager.ImageBitsPerPixel.ToString(), _
            m_TwainManager.ImagePixelType.ToString(), m_TwainManager.GetImageLayout().Left.ToString(), m_TwainManager.GetImageLayout().Top.ToString(), m_TwainManager.GetImageLayout().Right.ToString(), m_TwainManager.GetImageLayout().Bottom.ToString(), m_TwainManager.ImageLayoutDocumentNumber.ToString(), _
            m_TwainManager.ImageLayoutPageNumber.ToString(), m_TwainManager.ImageLayoutFrameNumber.ToString())
        Console.WriteLine("OnPostAllTransfer")
    End Sub

    Public Function OnPostTransfer(bit As Bitmap) As Boolean Implements IAcquireCallback.OnPostTransfer
        m_ImageCore.ImageBuffer.RemoveAllImages()
        m_ImageCore.IO.LoadImage(bit)
        Return True
    End Function

    Public Sub OnPreAllTransfers() Implements IAcquireCallback.OnPreAllTransfers
    End Sub

    Public Function OnPreTransfer() As Boolean Implements IAcquireCallback.OnPreTransfer
        Return True
    End Function

    Public Sub OnSourceUIClose() Implements IAcquireCallback.OnSourceUIClose
    End Sub

    Public Sub OnTransferCancelled() Implements IAcquireCallback.OnTransferCancelled
    End Sub

    Public Sub OnTransferError() Implements IAcquireCallback.OnTransferError
    End Sub

    Public Function GetAnnotations(iPageNumber As Integer) As Object Implements ISave.GetAnnotations
        If MultiPDF.Checked Then
            Return m_ImageCore.ImageBuffer.GetMetaData(CShort(iPageNumber), EnumMetaDataType.enumAnnotation)
        Else
            Return m_ImageCore.ImageBuffer.GetMetaData(m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer, EnumMetaDataType.enumAnnotation)
        End If
    End Function

    Public Function GetImage(iPageNumber As Integer) As Bitmap Implements ISave.GetImage
        If MultiPDF.Checked Then
            Return m_ImageCore.ImageBuffer.GetBitmap(CShort(iPageNumber))
        Else
            Return m_ImageCore.ImageBuffer.GetBitmap(m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer)
        End If
    End Function

    Public Function GetPageCount() As Integer Implements ISave.GetPageCount
        If MultiPDF.Checked Then
            Return m_ImageCore.ImageBuffer.HowManyImagesInBuffer
        Else
            Return 1
        End If
    End Function
    Private Sub Form1_FormClosed(sender As Object, e As FormClosedEventArgs)
        m_TwainManager.Dispose()
    End Sub
End Class

File: samples/C# Samples/ShowInfoAndSaveImage/ShowInfoAndSaveImage/Form1.cs
using Dynamsoft.Core;
using Dynamsoft.Core.Enums;
using Dynamsoft.TWAIN;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using Dynamsoft.TWAIN.Interface;
using Dynamsoft.PDF;
using System.IO;
using System.Runtime.InteropServices;

namespace ShowInfoAndSaveImage
{
    public partial class Form1 : Form,IAcquireCallback,ISave
    {
        EnumImageFileFormat sImageType;
        private TwainManager m_TwainManager = null;
        private ImageCore m_ImageCore = null;
        private Dynamsoft.PDF.PDFCreator m_PDFCreator = null;
        private String m_StrProductKey = "t0068MgAAAENENwNWc7+efmkY+t7se6XaRPFZkvfB7QWiTjHiLykxngQdY09pzVtOvrefXBbVvYFbJSluECHlyxaOvHwUADk=";
        public Form1()
        {
            InitializeComponent();
            m_TwainManager = new TwainManager(m_StrProductKey);
            m_ImageCore = new ImageCore();
            dsViewer1.Bind(m_ImageCore);
            dsViewer1.SetViewMode(1, 1);
            BMPradio.Checked = true;
            sImageType = 0;
            MultiTIFF.Enabled = false;
            MultiPDF.Enabled = false;
            m_PDFCreator = new PDFCreator(m_StrProductKey);
            m_RefreshInfo = new RefreshInfo(ShowImageInfo); 

        }


        private delegate void RefreshInfo(int strImageXResolution,int strImageYResolution,int strImageWidth,int strImageLength,string strImageBitsPerixel,string strImagePixelType,
                                        string strImageLayoutFrameLeft,string strImageLayoutFrameTop,string strImageLayoutFrameRight,string strImageLayoutFrameBottom,
                                        string strImageLayoutDocumentNumber,string strImageLayoutPageNumber,string strImageLayoutFrameNumber);


        private RefreshInfo m_RefreshInfo;

        public string GetImageSize()
        {
            string ImageSize = "";
            if (m_ImageCore.ImageBuffer.HowManyImagesInBuffer != 0)
            {
                ImageSize = m_ImageCore.IO.GetImageSizeWithSpecifiedType(m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer, sImageType).ToString();
            }
            return ImageSize;
        }

        private void btnScan_Click(object sender, EventArgs e)
        {

            List<string> tempList = new List<string>();
            for (int i = 0; i < m_TwainManager.SourceCount; i++)
            {
                tempList.Add(m_TwainManager.SourceNameItems((short)i));
            }
            SourceListWrapper tempSourceListWrapper = new SourceListWrapper(tempList);

            int iSelectIndex = tempSourceListWrapper.SelectSource();
            if (iSelectIndex == -1)
            {
                return;
            }
            else
            {
                m_TwainManager.IfDisableSourceAfterAcquire = true;
                m_TwainManager.IfShowUI = true;
                m_TwainManager.SelectSourceByIndex((short)iSelectIndex);
                m_TwainManager.AcquireImage(this as IAcquireCallback);
            }
        }

        private void TIFFradio_CheckedChanged(object sender, EventArgs e)
        {
            MultiTIFF.Enabled = true;
            MultiPDF.Enabled = false;
            sImageType = EnumImageFileFormat.WEBTW_TIF;
            txtFileSize.Text = GetImageSize();
        }

        private byte[] m_PDFFileBytes = null;
        private void PDFradio_CheckedChanged(object sender, EventArgs e)
        {
            MultiTIFF.Enabled = false;
            MultiPDF.Enabled = true;
            m_PDFFileBytes = null;
            m_PDFFileBytes = m_PDFCreator.SaveAsBytes(this as ISave);
            txtFileSize.Text = "0";
            if (m_PDFFileBytes != null)
            {
                txtFileSize.Text = m_PDFFileBytes.Length.ToString();
            }

        }

        private void BMPradio_CheckedChanged(object sender, EventArgs e)
        {
            MultiTIFF.Enabled = false;
            MultiPDF.Enabled = false;
            sImageType = EnumImageFileFormat.WEBTW_BMP;
            txtFileSize.Text = GetImageSize();

        }

        private void JPEGradio_CheckedChanged(object sender, EventArgs e)
        {
            MultiTIFF.Enabled = false;
            MultiPDF.Enabled = false;
            sImageType = EnumImageFileFormat.WEBTW_JPG;
            txtFileSize.Text = GetImageSize();
        }

        private void PNGradio_CheckedChanged(object sender, EventArgs e)
        {
            MultiTIFF.Enabled = false;
            MultiPDF.Enabled = false;
            sImageType = EnumImageFileFormat.WEBTW_PNG;
            txtFileSize.Text = GetImageSize();
        }

        private void BMPradio_MouseClick(object sender, MouseEventArgs e)
        {
            MultiTIFF.Enabled = false;
            MultiPDF.Enabled = false;
            sImageType = 0;
            txtFileSize.Text = GetImageSize();
        }

        private void btnSave_Click(object sender, EventArgs e)
        {
            try
            {
                if (m_ImageCore.ImageBuffer.HowManyImagesInBuffer > 0)
                {
                    string strFile = "";
                    if (BMPradio.Checked == true)
                    {
                        dlgFileSave.Filter = "BMP File (*.bmp)|*.bmp";
                    }
                    else if (JPEGradio.Checked == true)
                    {
                        dlgFileSave.Filter = "JPEG File (*.jpg)|*.jpg";
                    }
                    else if (PNGradio.Checked == true)
                    {
                        dlgFileSave.Filter = "PNG File (*.png)|*.png";
                    }
                    else if (TIFFradio.Checked == true)
                    {
                        dlgFileSave.Filter = "TIFF File (*.tif)|*.tif";
                    }
                    else if (PDFradio.Checked == true)
                    {
                        dlgFileSave.Filter = "PDF File (*.pdf)|*.pdf";
                    }
                    dlgFileSave.InitialDirectory = System.IO.Directory.GetCurrentDirectory();
                    dlgFileSave.FileName = "";
                    if (dlgFileSave.ShowDialog() == DialogResult.OK)
                    {
                        strFile = dlgFileSave.FileName;


                        if (BMPradio.Checked == true)
                        {
                            m_ImageCore.IO.SaveAsBMP(strFile, m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer);
                        }
                        else if (JPEGradio.Checked == true)
                        {

                            m_ImageCore.IO.SaveAsJPEG(strFile, m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer);
                        }
                        else if (PNGradio.Checked == true)
                        {

                            m_ImageCore.IO.SaveAsPNG(strFile, m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer);
                        }
                        else if (TIFFradio.Checked == true)
                        {
                            List<short> tempImageIndex = new List<short>();
                            if (MultiTIFF.Checked == true)
                            {
                                for (short sIndex = 0; sIndex < m_ImageCore.ImageBuffer.HowManyImagesInBuffer; sIndex++)
                                {
                                    tempImageIndex.Add(sIndex);
                                }
                            }
                            else
                            {
                                tempImageIndex.Add(m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer);
                            }

                            m_ImageCore.IO.SaveAsTIFF(strFile, tempImageIndex);
                        }
                        else if (PDFradio.Checked == true)
                        {
                            m_PDFCreator.Save(this as ISave, strFile);
                        }
                    }
                }
            }
            catch(Exception ex)
            {
                MessageBox.Show(ex.Message);
            }

            
        }

        private void btnRemove_Click(object sender, EventArgs e)
        {
            if (m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer != -1)
            {
                m_ImageCore.ImageBuffer.RemoveImage(m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer);
                btnScan.Enabled = true;
            }
        }

        private void Clear()
        {
            m_ImageCore.ImageBuffer.RemoveAllImages();
            edtXResolution.Text = "";
            edtYResolution.Text = "";
            edtWidth.Text = "";
            edtLength.Text = "";
            edtBitsPerPixel.Text = "";
            edtPixelType.Text = "";

            edtFrameLeft.Text = "";
            edtFrameTop.Text = "";
            edtFrameRight.Text = "";
            edtFrameBottom.Text = "";
            edtDocNumber.Text = "";
            edtPageNumber.Text = "";
            edtFrameNumber.Text = "";

            txtFileSize.Text = GetImageSize();
        }

        private void dynamicDotNetTwain_OnPostTransfer()
        {
            if (m_ImageCore.ImageBuffer.HowManyImagesInBuffer >= m_ImageCore.ImageBuffer.MaxImagesInBuffer)
            {
                btnScan.Enabled = false;
            }
        }

        private void ShowImageInfo(int strImageXResolution, int strImageYResolution, int strImageWidth, int strImageLength, string strImageBitsPerixel, string strImagePixelType, string strImageLayoutFrameLeft, string strImageLayoutFrameTop, string strImageLayoutFrameRight, string strImageLayoutFrameBottom,string strDocumentNumber,string strPageNumber,string strFrameNumber)
        {
            edtXResolution.Text = strImageXResolution.ToString();
            edtYResolution.Text = strImageYResolution.ToString();
            edtWidth.Text = strImageWidth.ToString();
            edtLength.Text = strImageLength.ToString();
            edtBitsPerPixel.Text = strImageLength.ToString();
            edtPixelType.Text = strImagePixelType.ToString();
            edtFrameLeft.Text = strImageLayoutFrameLeft.ToString();
            edtFrameTop.Text = strImageLayoutFrameTop.ToString();
            edtFrameRight.Text = strImageLayoutFrameRight.ToString();
            edtFrameBottom.Text = strImageLayoutFrameBottom.ToString();


            edtDocNumber.Text = strDocumentNumber;
            edtPageNumber.Text = strPageNumber;
            edtFrameNumber.Text = strFrameNumber;
            edtPageNumber.Text = m_TwainManager.ImageLayoutPageNumber.ToString();
            edtFrameNumber.Text = m_TwainManager.ImageLayoutFrameNumber.ToString();

            txtFileSize.Text = GetImageSize();
        }

        private void dynamicDotNetTwain_OnPreAllTransfers()
        {
            Clear();
        }

        public void LoadConvertResult(ConvertResult result)
        {
        }

        public void OnPostAllTransfers()
        {
            this.BeginInvoke(m_RefreshInfo, (int)m_TwainManager.ImageXResolution,
                            (int)m_TwainManager.ImageYResolution,
                            (int)m_TwainManager.ImageWidth,
                            (int)m_TwainManager.ImageLength,
                            m_TwainManager.ImageBitsPerPixel.ToString(),
                            m_TwainManager.ImagePixelType.ToString(),
                            m_TwainManager.GetImageLayout().Left.ToString(),
                            m_TwainManager.GetImageLayout().Top.ToString(),
                            m_TwainManager.GetImageLayout().Right.ToString(),
                            m_TwainManager.GetImageLayout().Bottom.ToString(),
                            m_TwainManager.ImageLayoutDocumentNumber.ToString(),
                            m_TwainManager.ImageLayoutPageNumber.ToString(),
                            m_TwainManager.ImageLayoutFrameNumber.ToString()
                            );
            Console.WriteLine("OnPostAllTransfer");
        }

        public bool OnPostTransfer(Bitmap bit)
        {
            m_ImageCore.ImageBuffer.RemoveAllImages();
            m_ImageCore.IO.LoadImage(bit);
            return true;
        }

        public void OnPreAllTransfers()
        {
        }

        public bool OnPreTransfer()
        {
            return true;
        }

        public void OnSourceUIClose()
        {
        }

        public void OnTransferCancelled()
        {
        }

        public void OnTransferError()
        {
        }

        public object GetAnnotations(int iPageNumber)
        {
            if (MultiPDF.Checked)
            {
                return m_ImageCore.ImageBuffer.GetMetaData((short)iPageNumber,EnumMetaDataType.enumAnnotation);
            }
            else
            {
                return m_ImageCore.ImageBuffer.GetMetaData(m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer,EnumMetaDataType.enumAnnotation);
            }
        }

        public Bitmap GetImage(int iPageNumber)
        {
            if (MultiPDF.Checked)
            {
                return m_ImageCore.ImageBuffer.GetBitmap((short)iPageNumber);
            }
            else
            {
                return m_ImageCore.ImageBuffer.GetBitmap(m_ImageCore.ImageBuffer.CurrentImageIndexInBuffer);
            }
        }

        public int GetPageCount()
        {
            if (MultiPDF.Checked)
            {
                return m_ImageCore.ImageBuffer.HowManyImagesInBuffer;
            }
            else
            {
                return 1;
            }
        }
        private void Form1_FormClosed(object sender, FormClosedEventArgs e)
        {
            m_TwainManager.Dispose();
        }
    }
}
